app/scripts/controllers/transactions/tx-state-manager.js

```diff
@@ -152,12 +152,15 @@ class TransactionStateManager extends EventEmitter {
     // or rejected tx's.
     // not tx's that are pending or unapproved
     if (txCount > txHistoryLimit - 1) {
-      const index = transactions.findIndex((metaTx) => {
-        return getFinalStates().includes(metaTx.status)
-      })
-      if (index !== -1) {
-        transactions.splice(index, 1)
-      }
+      do {
+        const index = transactions.findIndex((metaTx) => {
+          return getFinalStates().includes(metaTx.status)
+        })
+        if (index !== -1) {
+          transactions.splice(index, 1)
+        }
+        txCount = transactions.length
+      } while (txCount > txHistoryLimit - 1 && containsFinalStates(transactions))
```

Comment:
@Gudahtt does this satisfy your case?

Reply:
Well, if `index` was equal to `-1` (e.g. if `transactions.findIndex` didn't find anything) but `transactions.length` was still greater than `txHistoryLimit`, isn't it possible this could loop infinitely?

Maybe not if `containsFinalStates(transactions) === `transactions.some(tx => getFinalStates().includes(tx.status))`, but I wasn't sure whether that was guaranteed or not. If it is, then it's safe. But there shouldn't be any downside to adding a break, which'd make it more obvious that it can't loop infinitely.

Reply:
Oh nevermind, that is exactly what `containsFinalStates` does. This should be safe then - I didn't follow that the first time around for some reason.